janus_server/src/metrics.rs

```diff
@@ -0,0 +1,178 @@
+    #[serde(rename = "prometheus")]
+    Prometheus,
+    #[serde(rename = "otlp")]
+    Otlp(OtlpExporterConfiguration),
+}
+
+/// Configuration options specific to the OpenTelemetry OTLP metrics exporter.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OtlpExporterConfiguration {
+    /// gRPC endpoint for OTLP exporter.
+    pub endpoint: String,
+    /// Additional metadata/HTTP headers to be sent with OTLP requests.
+    #[serde(default)]
+    pub metadata: HashMap<String, String>,
+}
+
+/// Choice of OpenTelemetry metrics exporter implementation.
+pub enum MetricsExporterHandle {
+    #[cfg(feature = "prometheus")]
+    Prometheus(
+        opentelemetry_prometheus::PrometheusExporter,
+        tokio::task::JoinHandle<()>,
+    ),
+    #[cfg(feature = "otlp")]
+    Otlp(opentelemetry::sdk::metrics::controllers::PushController),
+    Noop,
+}
+
+#[derive(Debug)]
+struct CustomAggregatorSelector;
+
+impl AggregatorSelector for CustomAggregatorSelector {
+    fn aggregator_for(&self, descriptor: &Descriptor) -> Option<Arc<dyn Aggregator + Send + Sync>> {
+        match descriptor.instrument_kind() {
+            opentelemetry::metrics::InstrumentKind::ValueRecorder => {
+                // The following boundaries are copied from the Ruby and Go Prometheus clients.
+                // Buckets could be specialized per-metric by matching on `descriptor.name()`.
+                let boundaries = &[
+                    0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
+                ][..];
+                Some(Arc::new(
+                    opentelemetry::sdk::metrics::aggregators::histogram(descriptor, boundaries),
+                ))
+            }
+            opentelemetry::metrics::InstrumentKind::ValueObserver => Some(Arc::new(
+                opentelemetry::sdk::metrics::aggregators::last_value(),
+            )),
+            _ => Some(Arc::new(opentelemetry::sdk::metrics::aggregators::sum())),
+        }
+    }
+}
+
+/// Install a metrics provider and exporter, per the given configuration. The OpenTelemetry global
+/// API can be used to create and update meters, and they will be sent through this exporter. The
+/// returned handle should not be dropped until the application shuts down.
+pub fn install_metrics_exporter(
+    config: &MetricsConfiguration,
+) -> Result<MetricsExporterHandle, Error> {
+    match &config.exporter {
+        #[cfg(feature = "prometheus")]
+        Some(MetricsExporterConfiguration::Prometheus) => {
+            use http::StatusCode;
+            use hyper::Response;
+            use prometheus::{Encoder, TextEncoder};
+            use std::net::SocketAddr;
+            use warp::{Filter, Reply};
+
+            let exporter = opentelemetry_prometheus::exporter()
+                .with_aggregator_selector(CustomAggregatorSelector)
+                .try_init()?;
+
+            let filter = warp::path("metrics").and(warp::get()).map({
```

Comment:
What is the security story here -- how do we keep arbitrary users of the Internet from hitting this endpoint and receiving metric data? Will we be configuring things such that this metrics port is only exposed internally for our prometheus instance to scrape from? And if so -- how do we specify the port? I don't see anything in this code or in the Prometheus config that would allow the port to be controlled.

Reply:
Yes, we can keep the exporter port internal only. The exporter configuration builder takes the port from an environment variable or a default, I'll document that and add an option in our configuration.